feat(CommandPalette/InputMenu/Select/SelectMenu): unite object matching options

J-Michalek opened the pull request:
<!---
☝️ PR title should follow conventional commits (https://conventionalcommits.org)
-->

### 🔗 Linked issue

<!-- If it resolves an open issue, please link the issue here. For example "Resolves #123" -->

### ❓ Type of change

<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->

- [ ] 📖 Documentation (updates to the documentation or readme)
- [ ] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [ ] 👌 Enhancement (improving an existing functionality)
- [x] ✨ New feature (a non-breaking change that adds functionality)
- [ ] 🧹 Chore (updates to the build process or auxiliary tools and libraries)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description
I found it very useful to be able to compare objects by a certain key or by custom logic as comparison by reference is not always easily achievable.
<!-- Describe your changes in detail -->
<!-- Why is this change required? What problem does it solve? -->

### 📝 Checklist

<!-- Put an `x` in all the boxes that apply. -->
<!-- If your change requires a documentation PR, please link it appropriately -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] I have linked an issue or discussion.
- [ ] I have updated the documentation accordingly.

Commit:
feat(CommandPalette): add by prop for object matching

benjamincanac commented:
@J-Michalek This prop should also be implemented in Select, SelectMenu and InputMenu no? 🤔

J-Michalek commented:
@benjamincanac Hmm there may be more to this than I initially though. The `Select`, `SelectMenu` and `InputMenu` components already have `valueKey` prop which helps with the matching of objects by certain key, we can add the `by` prop to them as another option for object matching.

We could also add the `valueKey` to the `CommandPalette` for the sake of consistency, what do you think?

benjamincanac commented:
Indeed, adding a `value-key` would be better for consistency!

J-Michalek committed:
up

J-Michalek committed:
up

J-Michalek committed:
up

J-Michalek committed:
up

J-Michalek committed:
up

J-Michalek commented:
@benjamincanac I've added the `by` prop to `Select`, `SelectMenu` and `InputMenu`. I also added the `valueKey` prop to `CommandPalette`, but I got stuck on a TS issue where it seems that the modelValue/defaultValue somehow conflicts with `groups[number]['items']` type, but I could not figure out what is going on. When I tried to manually input the types into `CommandPaletteProps` the type looked correct, perhaps I am missing something regarding the generic types in a SFC?

benjamincanac commented on src/runtime/components/CommandPalette.vue:
```diff
@@ -180,13 +198,12 @@ import ULink from './Link.vue'
 import UKbd from './Kbd.vue'
 
-const props = withDefaults(defineProps<CommandPaletteProps<G, T>>(), {
-  modelValue: '',
```
I believe this was there for a reason and not to be removed 😬

sandros94 replied:
It does ring a bell but it also might be something that has been resolved with the types we have now

sandros94 commented:
Will happily take a look tomorrow, but in general I'm not a super fun of `by` naming. My first reaction before reading the description and comments on this PR was like "`by` what? is this an alias for `as`?". What about a more descriptive prop like `compareBy`?

benjamincanac commented:
The `by` props are overridden here to handle our types but it's inherited from Reka UI so I think keeping the same naming is the right move: https://github.com/unovue/reka-ui/blob/v2/packages/core/src/Listbox/ListboxRoot.vue#L22

sandros94 committed:
fix(CommandPalette): generic type order and defaults

sandros94 commented:
Pushed a commit mainly focused on ordering generics with their sequential use, but I'm noticing a strange `never` for `items` which breaks various things starting from `modelValue`

benjamincanac commented:
@sandros94 @J-Michalek What's left to do on this?

sandros94 commented:
> @sandros94 @J-Michalek What's left to do on this?

On my end a type improvement for the model value (which gets typed as never in some cases)

benjamincanac commented:
Maybe we should split the work of CommandPalette (`value-key` + generics) and `by` prop in two different PRs? 🤔

benjamincanac committed:
Merge branch 'v4' into pr/5158

benjamincanac committed:
up

benjamincanac commented:
I've merged all the conflicts but @sandros94 I think the generics are wrong, it should be `generic="T extends CommandPaletteItem[]"` right? We should also be using `DynamicSlots` I believe 🤔

sandros94 commented:
> I've merged all the conflicts but @sandros94 I think the generics are wrong, it should be `generic="T extends CommandPaletteItem[]"` right? We should also be using `DynamicSlots` I believe 🤔

Yes, but actually no 🥲. But yes, generics are wrong

`T` should actually be a union of all the groups, each with its own `CommandPaletteItem[]`. Other components only have nested arrays, which in typescript is far easier to handle, but in this case we have the nested array as part of a key inside an object that is part of an array. The issue I'm having (other than lack of time and work) is that I still need to objectively understand: _which came first, the chicken or the egg?_ (aka `CommandPaletteItem` or `CommandPaletteGroup`)

benjamincanac commented:
@J-Michalek There were interesting things in this PR but maybe we should split it into multiple PRs as it gets quite complex 😬

benjamincanac committed:
Merge branch 'v4' into pr/5158

benjamincanac committed:
up

benjamincanac committed:
Merge branch 'v4' into pr/5158

J-Michalek commented:
@benjamincanac Yeah, I will try to get a new PR going just for the addition of the `by` prop to the components and the more complex issue with types in `CommandPalette` would be handled in another PR.

benjamincanac commented:
Closing in favor of:
1. https://github.com/nuxt/ui/pull/5905
2. https://github.com/nuxt/ui/pull/5906

We're still having a generic issue in CommandPalette, #5905 only adds the `valueKey` support but this is tracked by: #2260
